### PR TITLE
Dt search

### LINKED ISSUE #9: Visitors must log in to access app
As a visitor
I cannot do anything except see welcome screen at "/"  with a login button
Once I log in
I am redirected to the root path
And I can search for recipes

### LINKED ISSUE #13: Filter Results
As a user
I can enter in my ingredients
and further filter my results by optional parameters (Dietary/Time/Type of meal)
My path is now "/recipes?keto=true&meal=dessert" <- need to think more about what we want the path to look like -> with optional params

### LINKED ISSUE #14: Sad Path Ingredients
As a user
When I enter in an invalid ingredient
I see a flash message that says "We couldn't find a recipe with your ingredients"
and it offers 10 random recipes

### LINKED ISSUE #18: Recipe Index Page
As a user
When I visit "/recipes"
I see up to 10 pictures with titles for 10 different recipes

Each recipe has title, number of missing ingredients, expandable summary

Each picture is a link to the recipe's show page

### PR DESCRIPTION
**Pull request by DavidTTran**
Completed:
- Recipe index pictures/titles are links to their show page
- Search functionality moved to root path, only available after a user logs in
- Sad paths for search function
   - No provided/conflicting params generates 10 random recipes
   - Provides up to 10 recipes

Refactor:
- Moved Spoonacular request/response into SpoonacularService
- Recipes are instantiated as a Recipe PORO
   - with `@id, @title, @image, @ingredients, @instructions, @time, @nutrition`

### CODE REVIEW

**Review by alex-latham — CHANGES_REQUESTED** [review]
Looking good, but the `require`s in RecipesController and poros/recipe.rb go against Rails' 'convention over configuration' philosophy. There are a couple ways to structure these POROs in a way that Rails will know about them without having a `require` -- the simplest being relocate the POROs folder to app/poros and then removing the `require(_relative)`s RecipesController and Recipe. Another alternative would be to rename the poros folder to spoonacular, and then wrapping the PORO classes inside in a Spoonacular module so you can call Spoonacular::Recipe from the RecipesController without the require. Wrapping them in a module also puts recipe_ingredients/instructions/nutrition into the same scope so the require_relatives would become unnecessary. Brownfield came with an example of the latter structure with its models/you_tube POROs folder/module.

 Sorry for the wall of text! I pushed up a branch agl_dt_search with a proof of concept if you are interested in taking a gander.

**Review by alex-latham — APPROVED** [review]

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+class RecipesController < ApplicationController
+  def index
+    recipes = SpoonacularService.new(search_params).parse_recipes
+    @recipes = recipes[:results].map do |recipe_json|
+      Recipe.new(recipe_json)
+    end
+    if @recipes.empty?
+      flash[:error] = "Sorry, we couldn't find any recipes matching your
```
Comment (mikez321) [inline app/controllers/recipes_controller.rb:9]: I like this part!  haha

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+class RecipesController < ApplicationController
+  def index
+    recipes = SpoonacularService.new(search_params).parse_recipes
```
Comment (mikez321) [inline app/controllers/recipes_controller.rb:4]: Probably a good place for a refactor later down the line.

File: app/poros/recipe.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+class Recipe
+  attr_reader :id,
```
Comment (mikez321) [inline app/poros/recipe.rb:2]: Are we using this ID for something later?  Do we need to expose it? - Nevermind
Oh... yeah, for the ID in the recipe show

File: app/services/spoonacular_service.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  def request_params
+    { apiKey: ENV['SPOONACULAR_KEY'],
+      addRecipeInformation: true,
+      includeIngredients: @params['ingredients'],
+      fillIngredients: true, ignorePantry: true,
+      maxReadyTime: @params['time'], diet: @params['diet'],
+      type: @params['type'],
+      minCalories: 0, minCholesterol: 0,
+      minFat: 0, minProtein: 0, minCarbs: 0, minSugar: 0, minSodium: 0,
+      number: 10, sort: 'random' }.compact
```
Comment (mikez321) [inline app/services/spoonacular_service.rb:22]: getting rid of nils, nice!

File: app/views/welcome/index.html.erb [hunk 5]
```diff
@@ -8,7 +8,41 @@
...
 </div>
-<% unless current_user %>
+
+<% if current_user %>
+  <%= form_tag "/search" do %>
+    <h2>INGREDIENTS</h2>
+      <%= text_field_tag :ingredients, nil, placeholder: "Enter your ingredients:" %><br>
+    <h2>DIET</h2>
+      <%= check_box_tag 'diet[]', 'vegetarian' %>
+      <%= label_tag :vegetarian %><br>
```
Comment (mikez321) [inline app/views/welcome/index.html.erb:18]: Just a heads up... I think we'll have pretty much all of these in a container of some sort and won't need the <br> tags.... I think...

**Review by alex-latham — APPROVED** [review]
Solid work. Everything looks good to me.

**Review by mikez321 — APPROVED** [review]
Looks great David!  Thanks for all your work on this and answering my questions along the way.

**Pull request closed without merging**